File: python/src/node.py
import numpy as np

class Node:
    def __init__(self,id,xyz,freedom):
        self.id = id
        self.xyz = xyz
        self.freedom = freedom
        self.dof = len(freedom)

    def print(self):
        print(self.id,":",self.xyz,",",self.freedom)

    def set_initial_condition(self):
        self.u  = np.zeros(self.dof,dtype=np.float64)
        self.um = np.zeros(self.dof,dtype=np.float64)
        self.v  = np.zeros(self.dof,dtype=np.float64)

        self.mass   = np.zeros(self.dof,dtype=np.float64)
        self.c      = np.zeros(self.dof,dtype=np.float64)
        self.k      = np.zeros(self.dof,dtype=np.float64)

        self.force = np.zeros(self.dof,dtype=np.float64)
        self.static_force = np.zeros(self.dof,dtype=np.float64)
        self.dynamic_force = np.zeros(self.dof,dtype=np.float64)

        self._up = np.zeros(self.dof,dtype=np.float64)
        self._ur = np.zeros(self.dof,dtype=np.float64)
        self._uy = np.zeros(self.dof,dtype=np.float64)

File: cpp/src/node.cpp
#include "all.h"
#include <Eigen/Core>
#include "node.h"

using EV = Eigen::VectorXd ;

Node::Node () {}
Node::Node (size_t id, std::vector<double> xyz, std::vector<size_t> freedom)
  {
    this->id = id;
    this->xyz = xyz;
    this->freedom = freedom;
    this->dof = freedom.size();
  }

void Node::print() {
  std::cout << this->id << ": ";
  for (size_t i = 0 ; i < this->xyz.size() ; i++) {
    std::cout << this->xyz.at(i) << " ";
  }
  std::cout << "-- ";
  for (size_t i = 0 ; i < this->freedom.size() ; i++) {
    std::cout << this->freedom.at(i) << " ";
  }
  std::cout << "\n";
}

void Node::set_initial_condition() {
  this->u = EV::Zero(this->dof);
  this->um = EV::Zero(this->dof);
  this->v  = EV::Zero(this->dof);

  this->mass = EV::Zero(this->dof);
  this->c    = EV::Zero(this->dof);
  this->k    = EV::Zero(this->dof);

  this->force = EV::Zero(this->dof);
  this->static_force  = EV::Zero(this->dof);
  this->dynamic_force = EV::Zero(this->dof);

  this->inv_mc = EV::Zero(this->dof);
  this->mass_inv_mc = EV::Zero(this->dof);
  this->c_inv_mc = EV::Zero(this->dof);
  this->dtdt_inv_mc = EV::Zero(this->dof);

  this->_up = EV::Zero(this->dof);
  this->_ur = EV::Zero(this->dof);
  this->_uy = EV::Zero(this->dof);

}
